### ISSUE TITLE
Different heads use different sparse masks.

### ISSUE THREAD

**Issue by helloyongyang** [SandAI-org/MagiAttention#160]
```
from magi_attention.functional import flex_flash_attn_func as ffa_func

q = torch.randn(seq_len, num_heads, head_dim, dtype=torch.bfloat16, device="cuda")

k = torch.randn(seq_len, num_heads, head_dim, dtype=torch.bfloat16, device="cuda")

v = torch.randn(seq_len, num_heads, head_dim, dtype=torch.bfloat16, device="cuda")


out = ffa_func(
    q,
    k,
    v,
    q_ranges=q_ranges,
    k_ranges=k_ranges,
    attn_type_map=attn_type_map,
    auto_range_merge=True,
)
```

Here, the q_ranges and k_ranges can be used to define a sparse block within the range of 0 to seq_len, for example:
```
q_ranges = torch.tensor([[  0, 128],
                              [  0, 128],
                              [128, 256]], device='cuda:0', dtype=torch.int32)

k_ranges = torch.tensor([[  0, 128],
                              [256, 306],
                              [128, 256]], device='cuda:0', dtype=torch.int32)
```

However, is it possible to set different q_ranges and k_ranges for different heads?
